### ISSUE TITLE
@QuarkusTest Error injecting io.quarkus.arc.runtime.ConfigStaticInitValues

### ISSUE THREAD

**Issue by rystidia** [quarkusio/quarkus#39521]
### Describe the bug

The Exception started appearing after migrating to Quarkus 3.8 from 3.2. It appears when several different @TestProfile() are used in a single run of @QuarkusTest. If a test has a different profile to the previously run test then Quarkus will be shut down and started with the new profile before running the tests. There is a GC notification coming asynchronously in a dedicated thread, and by the time it arrives, Quarkus CDI context is cleared, which results in NPE. The complete stacktrace looks like this:

```
java.lang.RuntimeException: Error injecting io.quarkus.arc.runtime.ConfigStaticInitValues io.quarkus.arc.runtime.ConfigStaticInitCheckInterceptor.configValues
	at io.quarkus.arc.runtime.ConfigStaticInitCheckInterceptor_Bean.doCreate(Unknown Source)
	at io.quarkus.arc.runtime.ConfigStaticInitCheckInterceptor_Bean.create(Unknown Source)
	at io.quarkus.arc.runtime.ConfigStaticInitCheckInterceptor_Bean.get(Unknown Source)
	at io.quarkus.arc.runtime.ConfigStaticInitCheckInterceptor_Bean.get(Unknown Source)
	at io.smallrye.config.inject.ConfigProducer_Subclass.<init>(Unknown Source)
	at io.smallrye.config.inject.ConfigProducer_Bean.doCreate(Unknown Source)
	at io.smallrye.config.inject.ConfigProducer_Bean.create(Unknown Source)
	at io.smallrye.config.inject.ConfigProducer_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:119)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:38)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
	at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.c19(Unknown Source)
	at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.computeIfAbsent(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:35)
	at io.quarkus.arc.impl.ClientProxies.getApplicationScopedDelegate(ClientProxies.java:21)
	at io.smallrye.config.inject.ConfigProducer_ClientProxy.arc$delegate(Unknown Source)
	at io.smallrye.config.inject.ConfigProducer_ClientProxy.arc_contextualInstance(Unknown Source)
	at io.smallrye.config.inject.ConfigProducer_ProducerMethod_produceBooleanConfigProperty_vanAxwNEeXT5HCx4jOWn3otsCog_Bean.doCreate(Unknown Source)
	at io.smallrye.config.inject.ConfigProducer_ProducerMethod_produceBooleanConfigProperty_vanAxwNEeXT5HCx4jOWn3otsCog_Bean.create(Unknown Source)
	at io.smallrye.config.inject.ConfigProducer_ProducerMethod_produceBooleanConfigProperty_vanAxwNEeXT5HCx4jOWn3otsCog_Bean.get(Unknown Source)
	at io.smallrye.config.inject.ConfigProducer_ProducerMethod_produceBooleanConfigProperty_vanAxwNEeXT5HCx4jOWn3otsCog_Bean.get(Unknown Source)
	at io.quarkus.arc.impl.CurrentInjectionPointProvider.get(CurrentInjectionPointProvider.java:48)
	at com.azul.cc.infrastructure.micrometer.DistributionStatisticConfigurator_Bean.doCreate(Unknown Source)
	at com.azul.cc.infrastructure.micrometer.DistributionStatisticConfigurator_Bean.create(Unknown Source)
	at com.azul.cc.infrastructure.micrometer.DistributionStatisticConfigurator_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:119)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:38)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
	at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.c39(Unknown Source)
	at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.computeIfAbsent(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:35)
	at io.quarkus.arc.impl.ClientProxies.getApplicationScopedDelegate(ClientProxies.java:21)
	at com.azul.cc.infrastructure.micrometer.DistributionStatisticConfigurator_ClientProxy.arc$delegate(Unknown Source)
	at com.azul.cc.infrastructure.micrometer.DistributionStatisticConfigurator_ClientProxy.map(Unknown Source)
	at io.micrometer.core.instrument.MeterRegistry.getMappedId(MeterRegistry.java:606)
	at io.micrometer.core.instrument.MeterRegistry.registerMeterIfNecessary(MeterRegistry.java:589)
	at io.micrometer.core.instrument.MeterRegistry.timer(MeterRegistry.java:336)
	at io.micrometer.core.instrument.Timer$Builder.register(Timer.java:459)
	at io.micrometer.core.instrument.Timer$Builder.register(Timer.java:453)
	at io.micrometer.core.instrument.binder.jvm.JvmGcMetrics$GcMetricsNotificationListener.handleNotification(JvmGcMetrics.java:211)
	at java.management/sun.management.NotificationEmitterSupport.sendNotification(NotificationEmitterSupport.java:155)
	at jdk.management/com.sun.management.internal.GarbageCollectorExtImpl.createGCNotification(GarbageCollectorExtImpl.java:115)
Caused by: java.lang.NullPointerException: Cannot invoke "io.quarkus.arc.ArcContainer.getActiveContext(java.lang.Class)" because "<local2>" is null
	at io.quarkus.arc.runtime.ConfigStaticInitValues_Bean.get(Unknown Source)
	at io.quarkus.arc.runtime.ConfigStaticInitValues_Bean.get(Unknown Source)
	... 42 more
```
Quarkus version 3.8.2
Grpc version 1.59.1
Protobuf version 3.25.0

### Output of `java -version`

openjdk version "17.0.4.1" 2022-08-12 LTS

### Quarkus version or git rev

3.8.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.0

**Comment by Ladicek** [comment]
The stack trace looks like metrics collection was not shut down properly. (Is that even possible with Micrometer? I don't know.)

**Comment by geoand** [comment]
Any chance you could provide a sample application that exhibits the problematic behavior?

**Comment by rystidia** [comment]
[This](https://github.com/rystidia/quarkus_npe_reproducer) reproducer seems to reproduce the error.

**Comment by mkouba** [comment]
So it seems that Micrometer attempts to use the `DistributionStatisticConfigurator` reference (which is registered as a global `MeterFilter`) after the tested quarkus app stopped. The `DistributionStatisticConfigurator` is a normal scoped bean and so a [client proxy is used](https://quarkus.io/guides/cdi#client_proxies) as a reference passed to the global registry. However, client proxies are not guaranteed to work outside the quarkus app. 

If you annotate your `MeterFilter` with [`@jakarta.inject.Singleton`](https://quarkus.io/guides/cdi#applicationscoped-and-singleton-look-very-similar-which-one-should-i-choose-for-my-quarkus-application) instead of `@jakarta.enterprise.context.ApplicationScoped` then no NPE is thrown because a direct reference is used instead (no client proxy).

In other words, it's not an ArC/CDI problem.

**Comment by rystidia** [comment]
It works, thank you so much. I am closing the issue

**Comment by geoand** [comment]
👍🏼

**Comment by Ladicek** [comment]
I actually think there _is_ a  bug: something (probably in the Micrometer extension?) doesn't shutdown properly. I don't know who would be the best person to ask, though. @brunobat maybe?

**Comment by brunobat** [comment]
This MeterFilter usage doesn't follow the recommendation on the Guide: https://quarkus.io/guides/telemetry-micrometer#meter-filter.
Here `@Singleton` is used. Maybe we need to be clearer on that regard.
Example:
```
    /** Define common tags that apply globally */
    @Produces
    @Singleton
    public MeterFilter configureAllRegistries() {
        return MeterFilter.commonTags(Arrays.asList(
                Tag.of("env", deploymentEnv)));
    }
```

**Comment by geoand** [comment]
Right, but I do agree with @Ladicek, that the fact that a proxy is or is not  used should not result in the problematic behavior.
If we have some, we should look into it a little more.

**Comment by Ladicek** [comment]
Yeah, using `@ApplicationScoped` just exposed a problem in that `io.micrometer.core.instrument.binder.jvm.JvmGcMetrics$GcMetricsNotificationListener.handleNotification()` keeps getting called even when the application is shutting down, which seems wrong. Maybe there's no way to unregister that listener, I don't know :shrug

**Comment by geoand** [comment]
I'll have a look when I have a little spare time

**Comment by geoand** [comment]
This one is actually pretty interested because the component that is causing the problem is the GC notification which happens when we re-augment Quarkus for the new test (because we have dropped the old CL).

So it's not an issue of the application shutdown phase.
